core/src/database/common.rs

```diff
@@ -697,6 +704,180 @@ impl Database {
+    pub async fn save_block_proof(
+        &self,
+        tx: Option<&mut sqlx::Transaction<'_, Postgres>>,
+        hash: block::BlockHash,
+        proof: Receipt,
+    ) -> Result<(), BridgeError> {
+        let proof = borsh::to_vec(&proof)?;
+
+        let query = sqlx::query("UPDATE header_chain_proofs SET proof = $1 WHERE block_hash = $2;")
+            .bind(proof)
+            .bind(BlockHashDB(hash));
+
+        match tx {
+            Some(tx) => query.execute(&mut **tx).await,
+            None => query.execute(&self.connection).await,
+        }?;
+
+        Ok(())
+    }
+
+    /// Gets a block's proof by referring to it by it's hash.
+    #[tracing::instrument(skip(self, tx), err(level = tracing::Level::ERROR), ret(level = tracing::Level::TRACE))]
+    pub async fn get_block_proof_by_hash(
+        &self,
+        tx: Option<&mut sqlx::Transaction<'_, Postgres>>,
+        hash: block::BlockHash,
+    ) -> Result<Option<Receipt>, BridgeError> {
+        let query = sqlx::query_as("SELECT proof FROM header_chain_proofs WHERE block_hash = $1;")
+            .bind(BlockHashDB(hash));
+
+        let receipt: (Option<Vec<u8>>,) = match tx {
+            Some(tx) => query.fetch_one(&mut **tx).await,
+            None => query.fetch_one(&self.connection).await,
+        }?;
+        let receipt = match receipt.0 {
+            Some(r) => r,
+            None => return Ok(None),
+        };
+
+        let receipt: Receipt = borsh::from_slice(&receipt)?;
+
+        Ok(Some(receipt))
+    }
+
+    /// Returns a block's hash and header, referring it to by it's height.
+    #[tracing::instrument(skip(self, tx), err(level = tracing::Level::ERROR), ret(level = tracing::Level::TRACE))]
+    pub async fn get_block_info_by_height(
+        &self,
+        tx: Option<&mut sqlx::Transaction<'_, Postgres>>,
+        height: u64,
+    ) -> Result<(block::BlockHash, block::Header), BridgeError> {
+        let query = sqlx::query_as(
+            "SELECT block_hash, block_header FROM header_chain_proofs WHERE height = $1;",
+        )
+        .bind(height as i64);
+
+        let result: (Option<BlockHashDB>, Option<BlockHeaderDB>) = match tx {
+            Some(tx) => query.fetch_one(&mut **tx).await,
+            None => query.fetch_one(&self.connection).await,
+        }?;
+
+        match result {
+            (Some(hash), Some(header)) => Ok((hash.0, header.0)),
+            _ => Ok((
+                // TODO: Do we need to return all zeroed values or an error?
+                BlockHash::all_zeros(),
+                Header {
+                    version: Version::TWO,
+                    prev_blockhash: BlockHash::all_zeros(),
+                    merkle_root: TxMerkleNode::all_zeros(),
+                    time: 0,
+                    bits: CompactTarget::default(),
+                    nonce: 0,
+                },
+            )),
+        }
+    }
+
+    /// Returns a block's hash and header, referring it to by it's height.
```

Comment:
Is this going to change? Right now we are using both height and hash for query.

Reply:
This is a recent addition for checking if a block in active blockchain exists in database or not (used in blockgazer::check_for_new_blocks). It's implementation details still has an ongoing discussion though (see https://github.com/chainwayxyz/clementine/pull/357#discussion_r1858333848).